### DISCUSSION TITLE
Idea: flexible boards and new FPGA supported

### DISCUSSION THREAD

**Discussion by Peter-van-Tol** [Peter-van-Tol/LiteX-CNC#112] [Ideas]
Was tinkering with my machine and was thinking:
1. desoldering the buffers on a 5A-75X is a lot of work, why not use a FPGA which does not have those buffers, whilst still cheap;
2. break-out boards, based on the DB25-connector, are quite rigid in their design, and may lead to some IO's not being used.

For the buffers I'm now looking into supporting Tang Nano 20K. First running from a Raspberry Pi over SPI (easy to port the code), later I want to support Ethernet via a W5500 chipset which translates SPI to UDP packages.

To make the break-out boards more flexible, I designed a new concept: a base plate on which different modules can be easily mounted. Each module, or leaf, can take 1 or more places on the base. For example:
- Isolated inputs (supports both NPN and PNP sensors) take 1 bay;
- Outputs (push-pull configuration, 1A max current) also take 1 bay;
- Encoder will take 2 bays without index or 3 bays with index;
- etc.

The board is designed to be also compatible with the HUB75HAT.

> **Costs**
> The standard size of a leaf will be approximately 15 mm x 45 mm, so the boards are very cheap to manufacture when delivered as a panel. Looking at €0.25 per board for the PCB; the carrier board will be €1.50 Euro for the PCB. So for the PCB only it will be total €5.50, excluding shipping and parts. The goal is to create a CNC controller for less then €50,--, including FPGA.

![image](https://github.com/user-attachments/assets/8faf708a-6c4c-4de7-8aed-13093c91b72d)

Wondering if this would be  good addition.

**Comment by kzali** [comment]
Have you looked at the Tang Primer 20K?  You may not need a W5500!

**Reply by hmnijp** [reply]
> looked at the Tang Primer 20K?

there is also colorlight i5 (or i9). this is a lattice ECP5U board without buffers, the form factor is similar to Tang Primer, but has 2x ethernet phy on the board, the motherboard will only require an ethernet connector with a transformer. In fact, you do not need to port the code for this board, the one for 5a-75 already works with it.

https://tomverbeure.github.io/2021/01/22/The-Colorlight-i5-as-FPGA-development-board.html

**Reply by Peter-van-Tol** [reply]
The ColorLight i5 (and i9) are supported indeed. Maybe have to alter the JSON-configurtation a bit to support them from the command-line. Although these are nice boards, they tend to be a little more expensive then the 5a-75X boards. The Tang Primer 20K is also more expensive than the TangNano 20K and the 5a-75X boards. So, that's the reason I'm not creating break-out boards for these boards.

When the boards have build-in PHY, they are already supported bij LiteX-CNC. When there is no PHY, and you need one, I'm looking into support that using the W5500. You'll be flexible in your design!

**Reply by kzali** [reply]
I think this would be a great idea.  I do have the Tang Primer but do not have the skill to port litex cnc considering that the dock has phy built in!
You're right the tangnano 20k and w5500 would be an economical solution.

**Reply by Peter-van-Tol** [reply]
> . In fact, you do not need to port the code for this board, the one for 5a-75 already works with it.

I created two more options in `main` to support the i5 and i9 directly from the config file:
```json
{
    "board_name": "simple_5a-75a",
    "board_type": "i5 v7.0",
    "clock_frequency": 50000000,
    "ethphy": {
        "tx_delay": 0
    },
   ...
```

```json
{
    "board_name": "simple_5a-75a",
    "board_type": "i9 v7.2",
    "clock_frequency": 50000000,
    "ethphy": {
        "tx_delay": 0
    },
   ...
```

The remainder of the config file can stay the same...
